tg/models.py

```diff
@@ -307,75 +309,60 @@ def prev_msg(self, chat_id: int, step: int = 1) -> bool:
         if new_idx < len(self.msgs[chat_id]):
             self.current_msgs[chat_id] = new_idx
             return True
-
         return False
 
     def get_message(self, chat_id: int, msg_id: int) -> Optional[Dict]:
-        msg_set = self.msg_ids[chat_id]
-        if msg_id not in msg_set:
-            # we are not storing any out of ordres old msgs
-            # just fetching then on demand
-            result = self.tg.get_message(chat_id, msg_id)
-            result.wait()
-            if result.error:
-                self.not_found.add(msg_id)
-                return None
-            return result.update
-        return next(iter(m for m in self.msgs[chat_id] if m["id"] == msg_id))
-
-    def remove_message(self, chat_id: int, msg_id: int):
-        msg_set = self.msg_ids[chat_id]
-        if msg_id not in msg_set:
-            return False
-        log.info(f"removing msg {msg_id=}")
-        # FIXME: potential bottleneck, replace with constan time operation
-        self.msgs[chat_id] = [
-            m for m in self.msgs[chat_id] if m["id"] != msg_id
-        ]
-        msg_set.remove(msg_id)
-        return True
+        with self.lock:
+            index = self.msg_ids[chat_id].get(msg_id)
+            if index:
+                return self.msgs[chat_id][index]
+        # we are not storing any out of ordres old msgs
+        # just fetching them on demand
+        result = self.tg.get_message(chat_id, msg_id)
+        result.wait()
+        if result.error:
+            self.not_found.add(msg_id)
+            return None
+        return result.update
+
+    def remove_messages(self, chat_id: int, msg_ids: List[int]):
+        with self.lock:
+            log.info(f"removing msg {msg_ids=}")
+            self.msgs[chat_id] = [
+                m for m in self.msgs[chat_id] if m["id"] not in msg_ids
+            ]
+            self.msg_ids[chat_id] = {
+                msg["id"]: i for i, msg in enumerate(self.msgs[chat_id])
+            }
+
+    def add_message(self, chat_id: int, msg: Dict[str, Any]):
+        with self.lock:
+            log.info(f"adding {msg=}")
+            self.msgs[chat_id].append(msg)
+            self.msgs[chat_id] = sorted(
```

Comment:
you can add condition here and sort only when `msg['id'] < self.msgs[chat_id][-1]['id']`

Reply:
also updating `msg_ids` could be simpler when `msg['id'] > self.msgs[chat_id][-1]['id'] == True` and it is a latest message. 
then you can do:
```
            self.msg_ids[chat_id][msg['id']] = self.msg_ids[chat_id][-1] + 1
```